### ISSUE TITLE
Allow for creation of a SpotifyClientAPI using a refresh token

### ISSUE THREAD

**Issue by jdpopkin** [adamint/spotify-web-api-kotlin#28]
Hi there! I'd like to be able to write my own code to fetch a refresh token from Spotify, then save that refresh token somewhere persistent, then have my app create a SpotifyClientAPI using that refresh token. Is there a way to do that?

I know that if I provide an auth code, this library will fetch its own refresh token. But if I rely on that, I need to go through the authentication process more or less every time I restart my app (unless my last restart was so recent that the old auth code is still valid). It'd be better if I could provide this library a refresh token manually.

(I'm new to Kotlin, by the way, so it's possible that there's a totally-obvious way to do this. I didn't see it in the main README, and I wasn't able to figure it out by looking around in SpotifyAPI.kt.)

Thanks for making this library!

**Comment by molikuner** [comment]
This is indeed possible. (version 2.0.0)
```Kotlin
val api = spotifyApi {
        credentials {
            clientId = "client id"
            clientSecret = "client secret"
            redirectUri = "callback uri"
        }
        clientAuthentication {
            token = Token(
                access_token = "a access token for example the first generated",
                token_type = "Bearer",
                expires_in = 0,
                refresh_token = "your refresh token",
                scope = arrayOf(
                    SpotifyScope.PLAYLIST_MODIFY_PRIVATE,
                    SpotifyScope.PLAYLIST_READ_PRIVATE,
                    SpotifyScope.USER_READ_PRIVATE
                ).joinToString(" ")
            )
        }
    }.buildClient(true)
```

The only problem with it currently is: #27

**Comment by adamint** [comment]
Hi @jdpopkin, not a problem!

Refresh tokens only last for a certain period of time as well (I think it's an hour, I can check). If you want to store it for later use, that may become an issue.

Could I ask which type of client you're trying to build? (SpotifyAPI vs SpotifyClientAPI)

**Comment by adamint** [comment]
@molikuner, that's currently only available for building `SpotifyClientAPI`. I'm changing that now

**Comment by adamint** [comment]
```kotlin
val refreshToken = spotifyApi {
        credentials {
            clientId = ""
            clientSecret=""
        }
    }.buildCredentialed().token


    val api = spotifyApi {
        credentials {
            clientId = ""
            clientSecret=""
        }
        authentication {
            token = refreshToken
        }
    }.buildCredentialed()
```

this is what you're looking for, right?

**Comment by adamint** [comment]
@jdpopkin could you retry with `2.0.1`?

**Comment by jdpopkin** [comment]
>Refresh tokens only last for a certain period of time as well (I think it's an hour, I can check). If you want to store it for later use, that may become an issue.

That's surprising to me. My understanding was that upon successful authorization, you receive both an access code (which directly allows access to user-specific Spotify API calls and only lasts a short time) and a refresh token (which allows you to get new access codes). I'm looking to create an app that lets me add new tracks to a specific playlist; if refresh tokens can't actually be persisted, it sounds like I would need the user to regularly re-authorize my app to modify their playlist, which seems strange.

>Could I ask which type of client you're trying to build? (SpotifyAPI vs SpotifyClientAPI)

I need to modify playlists, so I think I need SpotifyClientAPI.

>could you retry with 2.0.1? You now have the option to use the authentication block with an existing refresh token

Will do!

**Comment by adamint** [comment]
Hmm, interesting. You’re right - the refresh token is valid forever - sorry for saying otherwise! So just put the token in the authentication block and you should be good!

**Comment by adamint** [comment]
Can I resolve this?

**Comment by jdpopkin** [comment]
Yep! I haven't finished what I set out to do yet, but I can definitely confirm that I can create a `SpotifyClientAPI` using a refresh token on 2.0.1. Thank you!
